osquery/core/json.h

```diff
@@ -28,3 +37,55 @@
 // We need to reinclude this to re-enable boost's warning suppression
 #include <boost/config/compiler/visualc.hpp>
 #endif
+
+namespace osquery {
+/**
+ * @brief This provides a small wrapper around constructing JSON objects.
+ *
+ * Constructing RapidJSON objects can be slightly tricky. In our original
+ * refactoring we found several opportunities to leak memory and cause faults.
+ * This was mostly causes by setting allocators incorrectly.
+ */
+class JSON : private only_movable {
+ private:
+  explicit JSON(decltype(rapidjson::kObjectType) type);
+
+ public:
+  JSON(JSON&&) = default;
+  JSON& operator=(JSON&&) = default;
+
+  /// Create a JSON wrapper for an Object (map).
+  static JSON newObject();
+
+  /// Create a JSON wrapper for an Array (list).
+  static JSON newArray();
+
+ public:
+  /// Make a JSON object (map).
+  rapidjson::Document getObject() const;
+
+  /// Make a JSON array (list).
+  rapidjson::Document getArray() const;
+
+  /// Add a JSON object or array to a list.
+  void push(rapidjson::Document& line);
+
+  /// Add a string value to a JSON object.
+  void add(rapidjson::Document& line,
+           const std::string& key,
+           const std::string& value);
+
+  /// Add a size_t value to a JSON object.
+  void add(rapidjson::Document& line, const std::string& key, size_t value);
+
+  /// Add an int value to a JSON object.
+  void add(rapidjson::Document& line, const std::string& key, int value);
```

Comment:
What about add() for bool/double value? May be there are only couple of places they might be relevant!

Reply:
I'm still not married to this abstraction. :/

It is also limiting in that it only supports an array of objects. If the input document to the `add` methods was moved to the end and was an optional the default could be the managed document. Even though that works, it doesn't feel correct.